### PR TITLE
Convergence integration testing -- initial commit.

### PR DESCRIPTION
**Pull request by sam-falvo**
Getting the ball rolling.

This code was primarily written by @lvh while pair-prg'ing with @sam-falvo and @kivattik, so code review should be trivial.

### CODE REVIEW

File: otter/test/integration/utils.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    live_states = [ServerState.BUILD, ServerState.ACTIVE]
+    return len([s for s in servers if s.state in live_states])
+
+def _count_dead_servers(servers):
+    """
+    Count servers that are in error state.
+    """
+    return len([s for s in servers if s.state is ServerState.ERROR])
+
+def _sign(n):
```
Comment (lvh) [inline otter/test/integration/utils.py]: This function is no longer used.

File: otter/test/integration/test_utils.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
+"""
+Tests for the utility functions for convergence black-box testing.
+"""
+from pyrsistent import pmap, pset
+
+from twisted.trial.unittest import SynchronousTestCase
+
+from otter.convergence.model import (
```
Comment (radix) [inline otter/test/integration/test_utils.py]: Given that these are explicitly supposed to be "black-box" tests, I am surprised that internal model classes are being used.

**Comment by sam-falvo** [comment]
When this PR passes review, PLEASE DO NOT MERGE.  This will be a work-in-progress branch, and I'll create a sub-branch from this one for incremental reviews.

File: otter/test/integration/test_utils.py (hunk 2 above)
Comment (sam-falvo) [inline otter/test/integration/test_utils.py]: I don't recall what we were going to use these for.  I think for re-using any JSON marshaling in Otter?  @lvh can you confirm?

File: otter/test/integration/test_utils.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+        current_state = GroupState(
+            servers=self._create_servers(6),
+            lb_connections=pset([])
+        )
+        desired_state = DesiredGroupState(
+            launch_config=pmap(),
+            desired=5,
+        )
+        self.assertRaises(
+            AssertionError,
```
Comment (radix) [inline otter/test/integration/test_utils.py]: This AssertionError check could be a bit better; the function under test raises AssertionError in multiple different cases, and this isn't ensuring that it's doing it for the right reason. I'm not sure how important that is in this case, though.

**Comment by radix** [comment]
@sam-falvo If it doesn't cause the build to fail, I think it'd be better to just merge it to master as soon as it passes review. Maintaining a separate ephemeral master is just going to cause headaches. It's okay if you have to make changes to this code in later PRs.

File: otter/test/integration/test_utils.py (hunk 2 above)
Comment (lvh) [inline otter/test/integration/test_utils.py]: We're not doing any JSON marshalling.

Basically, we just needed some objects that describe all the state otter cares about, and re-using existing model classes is easier than writing new ones. These will never be the same instances as the functions in `otter.convergence` see, if that helps :)

**Comment by lvh** [comment]
@radix Agreed w.r.t regular merges of small PRs.

**Comment by cyli** [comment]
This LGTM - nice to have a just a rough metric to see if convergence is moving closer to its goal.

**Comment by sam-falvo** [comment]
@wirehead said it was OK to merge.

**Pull request merged**